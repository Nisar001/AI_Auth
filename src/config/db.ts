import 'dotenv/config';
import { DataSource } from 'typeorm';
import { User } from '../models/User';
import { OtpService } from '../models/OtpService';

export const AppDataSource = new DataSource({
  type: 'postgres',
  host: process.env.DB_HOST || 'localhost',
  port: parseInt(process.env.DB_PORT || '5432'),
  username: process.env.DB_USER || 'postgres',
  password: process.env.DB_PASS || '',
  database: process.env.DB_NAME || 'ai_auth_db',
  synchronize: true, // Set to false in production
  logging: false,
  entities: [User, OtpService],
  migrations: [],
  subscribers: [],
});
