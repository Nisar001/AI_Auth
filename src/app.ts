import 'reflect-metadata';
import express from 'express';
import cors from 'cors';
import dotenv from 'dotenv';
import { AppDataSource } from './config/db';
import { errorHandler, notFoundHandler } from './middlewares/errorHandler';
import { generalLimiter } from './middlewares/rateLimiter';
import authRoutes from './modules/routes/authRoutes';
import logger from './utils/logger';

// Load environment variables
dotenv.config();

const app = express();

// Security and parsing middleware
app.use(cors({
  origin: process.env.CORS_ORIGINS?.split(',') || ['http://localhost:3000'],
  credentials: true,
  methods: ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'],
  allowedHeaders: ['Content-Type', 'Authorization']
}));

app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true, limit: '10mb' }));

// Apply rate limiting to all requests
app.use(generalLimiter);

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({
    status: 'OK',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    environment: process.env.NODE_ENV || 'development'
  });
});

// API routes
app.use('/api/auth', authRoutes);

// Handle 404 errors
app.use(notFoundHandler);

// Global error handler
app.use(errorHandler);

// Initialize database connection
const initializeDatabase = async () => {
  try {
    await AppDataSource.initialize();
    logger.info('Database connected successfully');
  } catch (error) {
    logger.error('Database connection failed:', error);
    process.exit(1);
  }
};

// Graceful shutdown
process.on('SIGTERM', async () => {
  logger.info('SIGTERM received, shutting down gracefully');
  try {
    await AppDataSource.destroy();
    logger.info('Database connection closed');
    process.exit(0);
  } catch (error) {
    logger.error('Error during shutdown:', error);
    process.exit(1);
  }
});

export { app, initializeDatabase };