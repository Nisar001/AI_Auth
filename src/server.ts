import { app, initializeDatabase } from './app';
import logger from './utils/logger';

const PORT = parseInt(process.env.PORT || '3000', 10);

const startServer = async () => {
  try {
    // Initialize database connection
    await initializeDatabase();
    
    // Start the server
    const server = app.listen(PORT, () => {
      logger.info(`🚀 Server running on port ${PORT}`);
      logger.info(`📚 Environment: ${process.env.NODE_ENV || 'development'}`);
      logger.info(`🔗 Health check: http://localhost:${PORT}/health`);
      logger.info(`🔐 Auth API: http://localhost:${PORT}/api/auth`);
    });

    // Handle server errors
    server.on('error', (error: any) => {
      if (error.code === 'EADDRINUSE') {
        logger.error(`Port ${PORT} is already in use`);
      } else {
        logger.error('Server error:', error);
      }
      process.exit(1);
    });

    // Graceful shutdown on SIGINT
    process.on('SIGINT', () => {
      logger.info('SIGINT received, shutting down gracefully');
      server.close(() => {
        logger.info('Server closed');
        process.exit(0);
      });
    });

  } catch (error) {
    logger.error('Failed to start server:', error);
    process.exit(1);
  }
};

// Start the server
startServer();