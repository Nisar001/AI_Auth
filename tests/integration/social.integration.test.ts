// Mock Twilio and Nodemailer before any imports
jest.mock('twilio', () => {
  return jest.fn().mockImplementation(() => ({
    messages: { create: jest.fn().mockResolvedValue({ sid: 'mock-sid' }) }
  }));
});

jest.mock('nodemailer', () => {
  return {
    createTransport: jest.fn(() => ({
      sendMail: jest.fn().mockResolvedValue({ messageId: 'mock-message-id' }),
      verify: jest.fn().mockResolvedValue(true)
    }))
  };
});

jest.mock('../../src/config/db', () => {
  return {
    AppDataSource: {
      getRepository: jest.fn(() => {
        return {
          create: jest.fn((entity: any) => {
            return { id: 1, ...entity };
          }),
          save: jest.fn(async (entity: any) => {
            return { id: 1, ...entity };
          }),
          findOne: jest.fn(async (query: any) => {
            if (query.where.socialId === 'social123') {
              return { id: 1, email: 'social@example.com', socialId: 'social123' };
            }
            return null;
          }),
        };
      }),
      initialize: jest.fn().mockResolvedValue(undefined),
      synchronize: jest.fn().mockResolvedValue(undefined),
      destroy: jest.fn().mockResolvedValue(undefined),
    },
  };
});

import request from 'supertest';
import { app } from '../../src/app';
import { AppDataSource } from '../../src/config/db';
import dotenv from 'dotenv';

dotenv.config({ path: '.env.test' });

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Social Login Integration', () => {
  it('should get Google OAuth URL', async () => {
    const res = await request(app)
      .get('/api/auth/google/auth-url');

    expect([200, 201, 400]).toContain(res.status);
    expect(res.body).toMatchSnapshot();
  });

  it('should get GitHub OAuth URL', async () => {
    const res = await request(app)
      .get('/api/auth/github/auth-url');

    expect([200, 201, 400]).toContain(res.status);
    expect(res.body).toMatchSnapshot();
  });

  // Refactor the social login test
  it('should handle social login successfully', async () => {
    const repository = AppDataSource.getRepository('User');

    // Mock repository methods
    jest.spyOn(repository, 'create').mockImplementation((entity: any) => {
      return { id: 1, ...entity };
    });

    jest.spyOn(repository, 'save').mockImplementation(async (entity: any) => {
      return { id: 1, ...entity };
    });

    jest.spyOn(repository, 'findOne').mockImplementation(async (query: any) => {
      if (query.where.socialId === 'social123') {
        return { id: 1, email: 'social@example.com', socialId: 'social123' };
      }
      return null;
    });

    // Test logic
    const user = repository.create({ email: 'social@example.com', password: 'password123', socialId: 'social123' });
    await repository.save(user);

    const savedUser = await repository.findOne({ where: { socialId: 'social123' } });
    expect(savedUser).not.toBeNull();
    expect(savedUser!.email).toBe('social@example.com');
    expect(savedUser).toMatchSnapshot({
      id: expect.any(Number),
      email: expect.any(String),
      socialId: expect.any(String),
    });
  });
});
