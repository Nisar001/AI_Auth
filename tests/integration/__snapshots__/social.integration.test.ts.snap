// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social Login Integration should get GitHub OAuth URL 1`] = `
{
  "data": {
    "authUrl": "https://github.com/login/oauth/authorize?client_id=Ov23li0YXXoBXZQpDI23&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fauth%2Fgithub%2Fcallback&scope=user%3Aemail&state=f03dbc2a9144c9fb36ffdd874ecb95edd21c41bec2c1c31384f329fd2f833f0a",
    "provider": "github",
    "state": "f03dbc2a9144c9fb36ffdd874ecb95edd21c41bec2c1c31384f329fd2f833f0a",
  },
  "message": "GitHub OAuth URL generated successfully",
  "statusCode": 200,
  "success": true,
}
`;

exports[`Social Login Integration should get Google OAuth URL 1`] = `
{
  "data": {
    "authUrl": "https://accounts.google.com/o/oauth2/v2/auth?client_id=1088061610840-prcu4cbids3vkli59ns903pa2ko4t9j8.apps.googleusercontent.com&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fauth%2Fgoogle%2Fcallback&response_type=code&scope=openid%20email%20profile&state=3e54528b99bf8ba836db0da828dab6f4c40f22e381af6df99fb3783cd956cc67&access_type=offline&prompt=consent",
    "provider": "google",
    "state": "3e54528b99bf8ba836db0da828dab6f4c40f22e381af6df99fb3783cd956cc67",
  },
  "message": "Google OAuth URL generated successfully",
  "statusCode": 200,
  "success": true,
}
`;

exports[`Social Login Integration should handle social login successfully 1`] = `
{
  "email": Any<String>,
  "id": Any<Number>,
  "socialId": Any<String>,
}
`;
